libraries/rush-lib/src/logic/operations/ShardedPhaseOperationPlugin.ts

```diff
@@ -0,0 +1,85 @@
+// See LICENSE in the project root for license information.
+
+import type { ITerminal } from '@rushstack/terminal';
+import type { IPhase } from '../../api/CommandLineConfiguration';
+import type { RushConfigurationProject } from '../../api/RushConfigurationProject';
+import type {
+  ICreateOperationsContext,
+  IPhasedCommandPlugin,
+  PhasedCommandHooks
+} from '../../pluginFramework/PhasedCommandHooks';
+import { Operation } from './Operation';
+import { type IRushPhaseSharding, RushProjectConfiguration } from '../../api/RushProjectConfiguration';
+
+export const PLUGIN_NAME: 'ShardedPhasedOperationPlugin' = 'ShardedPhasedOperationPlugin';
+
+/**
+ * Phased command that
+ */
+export class ShardedPhasedOperationPlugin implements IPhasedCommandPlugin {
+  public apply(hooks: PhasedCommandHooks): void {
+    hooks.createOperations.tapPromise(PLUGIN_NAME, spliceShards);
+  }
+}
+
+async function spliceShards(
+  existingOperations: Set<Operation>,
+  context: ICreateOperationsContext
+): Promise<Set<Operation>> {
+  const { terminal } = context;
+
+  const shardableOperations: Set<Operation> = new Set<Operation>();
+
+  for (const operation of existingOperations) {
+    const { associatedPhase: phase, associatedProject: project } = operation;
+    const shardConfig: IRushPhaseSharding | undefined = await getShardConfig(phase!, project!, terminal!);
+    if (shardConfig) {
+      shardableOperations.add(operation);
+    }
+  }
+
+  for (const operation of shardableOperations) {
+    const { associatedPhase: phase, associatedProject: project } = operation;
+    const shardConfig: IRushPhaseSharding | undefined = await getShardConfig(phase!, project!, terminal!);
+    if (phase && project && shardConfig && shardConfig.count > 1) {
+      const shards: number = shardConfig.count;
+      existingOperations.delete(operation);
+      for (const shard of Array.from({ length: shards }, (element, index) => index + 1)) {
+        const shardOperation: Operation = new Operation({
+          project,
+          phase,
+          shard: {
+            current: shard,
+            total: shards
+          },
+          runner: operation.runner
+        });
+
+        for (const dependency of operation.dependencies) {
+          shardOperation.addDependency(dependency);
+          operation.deleteDependency(dependency);
+        }
+        for (const dependent of operation.consumers) {
+          dependent.addDependency(shardOperation);
+          dependent.deleteDependency(operation);
+        }
+        existingOperations.add(shardOperation);
+      }
+    }
+  }
+
+  return existingOperations;
+}
+
+async function getShardConfig(
+  phase: IPhase,
+  project: RushConfigurationProject,
+  terminal: ITerminal
+): Promise<IRushPhaseSharding | undefined> {
+  const rushProjectConfiguration: RushProjectConfiguration | undefined =
+    await RushProjectConfiguration.tryLoadForProjectAsync(project, terminal);
```

Comment:
This is available as `context.projectConfigurations.get(project)`, so you don't need anything async.

Reply:
Updated in [7d34e74](https://github.com/microsoft/rushstack/pull/4652/commits/7d34e74bc94de34fb1bafe0c89cde48b7d01c7de)